### PR TITLE
Move xfem tests 2

### LINKED ISSUE #6757: Move XFEM tests
### Description of the enhancement or error report
We originally put a number of XFEM tests that depend on solid_mechanics in the combined module. Since the xfem module depends on solid_mechanics, we can move them back into xfem.

### Rationale for the enhancement or information for reproducing the error
Better to have tests in the appropriate module if possible.

### Identified impact
Move modules/combined/tests/solid_mechanics/xfem directory to xfem/tests.

### PR DESCRIPTION
**Pull request by bwspenc**
closes #6757

### CODE REVIEW

**Comment by bwspenc** [comment]
@friedmud After thinking this over a little more, I think this is a far better solution than my previous proposal. It doesn't involve any use of `static`.  This just calls `registerObjects()` for the modules that are depended upon from the module app's constructor rather than from its `registerObjects()`. The only downside I see is that when there is a chain of dependencies (as we have with xfem->solid_mechanics->tensor_mechanics), you have to register the objects for all of them in the module that depends on them. This doesn't break MultiApps like the other version did.  There's a test in phase_field that tests that.

**Comment by bwspenc** [comment]
@dschwen What do you think of this?  I'm not tied to this solution, but this seems as easy as it could get, and it works.

File: modules/phase_field/src/base/PhaseFieldApp.C [hunk 1]
```diff
@@ -218,9 +219,11 @@ PhaseFieldApp::PhaseFieldApp(const InputParameters & parameters) :
     MooseApp(parameters)
 {
   Moose::registerObjects(_factory);
+  TensorMechanicsApp::registerObjects(_factory);
   PhaseFieldApp::registerObjects(_factory);
 
   Moose::associateSyntax(_syntax, _action_factory);
+  TensorMechanicsApp::associateSyntax(_syntax, _action_factory);
```
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: PhaseField does not depend on TM anymore

Comment (bwspenc) [inline modules/phase_field/src/base/PhaseFieldApp.C]: Are you sure?  It says `TENSOR_MECHANICS := yes` in its makefile.  Maybe that needs to be changed.

**Comment by dschwen** [comment]
I'm sorry, but I have to be all negative about this again. Module interdependencies should be _opaque_ to application developers. 

What you are doing here is introducing potentially wide sweeping changes to object registration that will affect _every_ application developer, only to _not_ use a capability that we _alrady have_ that was developed and implemented for _exactly this purpose_!

File: modules/phase_field/src/base/PhaseFieldApp.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: Uhm, I has a PR a week ago that removed that (where I moved the tensors into framework). Let me check.

Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: Yeah, looks like I forgot to remove that entry (I removed the automatic dependence in modules.mk, but forgot the explicit one)

**Comment by bwspenc** [comment]
@dschwen If a module depends on another module, it seems reasonable that there would be a small amount of code required to bring in those capabilities into the app that gets built for that module. This just mimics the way the applications bring in modules for their constructors. This doesn't affect applications that depend on these modules at all because they don't create instances of ModuleApp, so they never even execute this code. 

Sure, we have a system for dynamically loading modules, but there's the hassle the user has to deal with of telling it where to find them.  Until we have some way of locating those automatically, that would have a far larger impact on users than this.

**Comment by dschwen** [comment]
But for the tests we know where they are. They are always in the same location.

**Comment by bwspenc** [comment]
...unless you copy the test file somewhere else, modify it, and try to run it (which I do all the time).

**Comment by friedmud** [comment]
@dschwen While I agree with you that I don't want to modify the registration system just to help with this one "special case"... I cannot agree with using dynamic registration here.

For one thing... dynamic registration does NOT inform the build system to "go build" the other module.  So what happens if the other module is not built when you try to run tests?  They are just going to fail?  That would suck.

I say we table this entire discussion until @permcody gets back.  He is in charge (both of MOOSE and the registration system) and invented the dynamic registration system.  I would like to hear what he has to say.

**Comment by dschwen** [comment]
Agreed on the build system point.

**Comment by bwspenc** [comment]
I think it's important to emphasize that this change doesn't modify the registration system at all.  It just makes the module's app use it the same way an app uses it.  This is identical to what is done in AnimalApp.C. Since the module's executable doesn't affect another app that uses it as a library, this has no impact on that app that depends on it.
I agree that this discussion can wait until @permcody gets back.  This isn't urgent at all.

**Comment by dschwen** [comment]
I think I may have misunderstood the scope of the change! This doesn't look so horrible anymore.

**Comment by friedmud** [comment]
Yes - it looks ok now - but I would still like to hear @permcody 's opinion.

**Comment by permcody** [comment]
This new proposal is the correct way to do this. Each application needs to bring in the modules it depends in by name which I know seems a little nasty. There isn't a clean way of doing this automatically based on the Makefile alone if we want to continue to support static builds. I've been down this path before and it turns into a huge mess trying to dynamically define objects and have empty bodies and empty headers all over to support static linkage. Now the day we decide to drop that support, we could come up with a much more elegant system. 😉

**Comment by permcody** [comment]
👍

**Pull request merged**